// @microsoft/dev-tunnels-connections/refreshingTunnelEventArgs.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.RefreshingTunnelEventArgs = void 0;
/**
 * Event args for tunnel refresh event.
 */
class RefreshingTunnelEventArgs {
    /**
     * Creates a new instance of RefreshingTunnelAccessTokenEventArgs class.
     */
    constructor(
    /**
     * Tunnel access scope to get the token for.
     */
    tunnelAccessScope, 
    /**
     * Tunnel being refreshed.
     */
    tunnel, 
    /**
     * A value indicating whether ports need to be included into the refreshed tunnel.
     */
    includePorts, 
    /**
     * Management client used for connections.
     */
    managementClient, 
    /**
     * Cancellation token that event handler may observe when it asynchronously fetches the tunnel access token.
     */
    cancellation) {
        this.tunnelAccessScope = tunnelAccessScope;
        this.tunnel = tunnel;
        this.includePorts = includePorts;
        this.managementClient = managementClient;
        this.cancellation = cancellation;
    }
}
exports.RefreshingTunnelEventArgs = RefreshingTunnelEventArgs;
