// @microsoft/dev-tunnels-ssh/trace.js
"use strict";
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//
Object.defineProperty(exports, "__esModule", { value: true });
exports.SshTraceEventIds = exports.TraceLevel = void 0;
/**
 * SSH trace event level.
 */
var TraceLevel;
(function (TraceLevel) {
    TraceLevel["Error"] = "error";
    TraceLevel["Warning"] = "warning";
    TraceLevel["Info"] = "info";
    TraceLevel["Verbose"] = "verbose";
})(TraceLevel = exports.TraceLevel || (exports.TraceLevel = {}));
// Events defined below should stay in sync with those in C#.
const baseEventId = 9000;
class SshTraceEventIds {
}
exports.SshTraceEventIds = SshTraceEventIds;
// Error / Warning events
SshTraceEventIds.unknownError = baseEventId + 0;
SshTraceEventIds.streamReadError = baseEventId + 1;
SshTraceEventIds.streamWriteError = baseEventId + 2;
SshTraceEventIds.streamCloseError = baseEventId + 3;
SshTraceEventIds.sendMessageFailed = baseEventId + 4;
SshTraceEventIds.receiveMessageFailed = baseEventId + 5;
SshTraceEventIds.handleMessageFailed = baseEventId + 6;
SshTraceEventIds.serverAuthenticationFailed = baseEventId + 7;
SshTraceEventIds.clientAuthenticationFailed = baseEventId + 8;
SshTraceEventIds.authenticationError = baseEventId + 9;
SshTraceEventIds.channelWindowAdjustFailed = baseEventId + 10;
SshTraceEventIds.channelWaitForWindowAdjust = baseEventId + 11;
SshTraceEventIds.sessionReconnectInitFailed = baseEventId + 20;
SshTraceEventIds.serverSessionReconnectFailed = baseEventId + 21;
SshTraceEventIds.clientSessionReconnectFailed = baseEventId + 22;
SshTraceEventIds.sessionRequestFailed = baseEventId + 23;
SshTraceEventIds.channelRequestFailed = baseEventId + 24;
SshTraceEventIds.serverListenFailed = baseEventId + 50;
SshTraceEventIds.portForwardServerListenFailed = baseEventId + 51;
SshTraceEventIds.portForwardRequestInvalid = baseEventId + 52;
SshTraceEventIds.portForwardChannelInvalid = baseEventId + 53;
SshTraceEventIds.portForwardChannelOpenFailed = baseEventId + 54;
SshTraceEventIds.portForwardConnectionFailed = baseEventId + 55;
SshTraceEventIds.metricsError = baseEventId + 61;
SshTraceEventIds.keepAliveFailed = baseEventId + 62;
SshTraceEventIds.keepAliveRequestReceived = baseEventId + 63;
SshTraceEventIds.keepAliveResponseNotReceived = baseEventId + 64;
// Info / Verbose events
SshTraceEventIds.protocolVersion = baseEventId + 100;
SshTraceEventIds.sendingMessage = baseEventId + 101;
SshTraceEventIds.receivingMessage = baseEventId + 102;
SshTraceEventIds.sendingChannelData = baseEventId + 103;
SshTraceEventIds.receivingChannelData = baseEventId + 104;
SshTraceEventIds.sessionEncrypted = baseEventId + 110;
SshTraceEventIds.sessionAuthenticating = baseEventId + 111;
SshTraceEventIds.sessionAuthenticated = baseEventId + 112;
SshTraceEventIds.sessionClosing = baseEventId + 113;
SshTraceEventIds.sessionConnecting = baseEventId + 114;
SshTraceEventIds.channelOpened = baseEventId + 120;
SshTraceEventIds.channelOpenFailed = baseEventId + 121;
SshTraceEventIds.channelEofReceived = baseEventId + 122;
SshTraceEventIds.channelClosed = baseEventId + 123;
SshTraceEventIds.serverListening = baseEventId + 150;
SshTraceEventIds.serverClientConnected = baseEventId + 151;
SshTraceEventIds.portForwardServerListening = baseEventId + 152;
SshTraceEventIds.portForwardConnectionAccepted = baseEventId + 153;
SshTraceEventIds.portForwardChannelOpened = baseEventId + 154;
SshTraceEventIds.portForwardChannelClosed = baseEventId + 155;
SshTraceEventIds.portForwardConnectionOpened = baseEventId + 156;
SshTraceEventIds.portForwardConnectionClosed = baseEventId + 157;
SshTraceEventIds.sessionDisconnected = baseEventId + 160;
SshTraceEventIds.clientSessionReconnecting = baseEventId + 161;
SshTraceEventIds.serverSessionReconnecting = baseEventId + 162;
SshTraceEventIds.clientSessionStartReconnecting = baseEventId + 163;
SshTraceEventIds.algorithmNegotiation = baseEventId + 170;
SshTraceEventIds.debugMessage = baseEventId + 200;
